src/tpm2-tss-engine-ciphers.c

```diff
@@ -0,0 +1,393 @@
+/*******************************************************************************
+ * Copyright 2019, Schneider-Electric
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of tpm2-tss-engine nor the names of its contributors
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#include <string.h>
+#include <endian.h>
+
+#include <openssl/engine.h>
+#include <openssl/evp.h>
+#include <tss2/tss2_esys.h>
+
+#include "tpm2-tss-engine.h"
+#include "tpm2-tss-engine-common.h"
+
+#define HANDLE_SIZE 8
+
+static int tpm2_cipher_nids[] = {
+    NID_aes_256_cfb128,
+    NID_aes_256_ofb128,
+    0
+};
+
+static int
+big_endian_array_to_native(const unsigned char *in, size_t size)
+{
+    uint32_t r;
+
+    memcpy(&r, in, size);
+    r = be32toh(r);
+
+    return r;
+}
+
+static int
+populate_tpm2data(const unsigned char *key, TPM2_DATA **tpm2Data)
+{
+    uint32_t keyHandle = 0;
+
+    /* Use an empty Key */
+    if (key == NULL) {
+        *tpm2Data = NULL;
+
+    /* Read persistent key, use for openssl app : openssl enc */
+    } else if ((key[0] == 0x81) && (key[HANDLE_SIZE/2 + 1] == 0)) {
```

Comment:
That not possible here. `strnlen(key, HANDLE_SIZE*2+1)` will always return 1 because the second byte of data is `0x00` in a key handle, for example `0x81000001`

Reply:
Oh sorry, of course you are right.